flashlight/app/imgclass/examples/ImageNetTransformer.cpp

```diff
@@ -0,0 +1,325 @@
+
+using namespace fl;
+using fl::ext::image::compose;
+using fl::ext::image::ImageTransform;
+using namespace fl::app::imgclass;
+
+#define FL_LOG_MASTER(lvl) LOG_IF(lvl, (fl::getWorldRank() == 0))
+
+const double pi = std::acos(-1);
+
+// Returns the average loss, top 5 error, and top 1 error
+std::tuple<double, double, double> evalLoop(
+    std::shared_ptr<Container> model,
+    Dataset& dataset) {
+  AverageValueMeter lossMeter;
+  TopKMeter top5Acc(5);
+  TopKMeter top1Acc(1);
+
+  // Place the model in eval mode.
+  model->eval();
+  for (auto& example : dataset) {
+    auto inputs = noGrad(example[kImagenetInputIdx]);
+    auto output = model->forward({inputs}).front();
+
+    auto target = noGrad(example[kImagenetTargetIdx]);
+
+    // Compute and record the loss.
+    auto loss = categoricalCrossEntropy(output, target);
+    lossMeter.add(loss.array().scalar<float>());
+    top5Acc.add(output.array(), target.array());
+    top1Acc.add(output.array(), target.array());
+  }
+  model->train();
+  fl::ext::syncMeter(lossMeter);
+  fl::ext::syncMeter(top5Acc);
+  fl::ext::syncMeter(top1Acc);
+
+  double loss = lossMeter.value()[0];
+  return std::make_tuple(loss, top5Acc.value(), top1Acc.value());
+};
+
+int main(int argc, char** argv) {
+  fl::init();
+  google::InitGoogleLogging(argv[0]);
+  google::InstallFailureSignalHandler();
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+
+  const std::string labelPath = lib::pathsConcat(FLAGS_data_dir, "labels.txt");
+  const std::string trainList = lib::pathsConcat(FLAGS_data_dir, "train");
+  const std::string valList = lib::pathsConcat(FLAGS_data_dir, "val");
+
+  /////////////////////////
+  // Setup distributed training
+  ////////////////////////
+  af::info();
+  if (FLAGS_distributed_enable) {
+    fl::ext::initDistributed(
+        FLAGS_distributed_world_rank,
+        FLAGS_distributed_world_size,
+        FLAGS_distributed_max_devices_per_node,
+        FLAGS_distributed_rndv_filepath);
+  }
+  const int worldRank = fl::getWorldRank();
+  const int worldSize = fl::getWorldSize();
+  const bool isMaster = (worldRank == 0);
+
+  af::setDevice(worldRank);
+  af::setSeed(worldSize);
+
+  auto reducer =
+      std::make_shared<fl::CoalescingReducer>(1.0 / worldSize, true, true);
+
+  //////////////////////////
+  //  Create datasets
+  /////////////////////////
+  FL_LOG_MASTER(INFO) << "Creating dataset";
+  // These are the mean and std for each channel of Imagenet
+  const std::vector<float> mean = {0.485, 0.456, 0.406};
+  const std::vector<float> std = {0.229, 0.224, 0.225};
+  const int randomResizeMax = 480;
+  const int randomResizeMin = 256;
+  const int randomCropSize = 224;
+  const float horizontalFlipProb = 0.5f;
+  // TransformDataset will apply each transform in a vector to the respective
+  // af::array. Thus, we need to `compose` all of the transforms so are each
+  // applied only to the image
+  ImageTransform trainTransforms = compose(
```

Comment:
here they use a bit different policy https://github.com/google-research/vision_transformer/blob/00883dd691c63a6830751563748663526e811cee/vit_jax/input_pipeline.py#L118-L130 (inception_crop=True in their training https://github.com/google-research/vision_transformer/blob/master/vit_jax/train.py#L84)